docs/Migration Guide.md

```diff
@@ -5,11 +5,78 @@ of SpiderMonkey to the next ESR version.
 
 ## ESR 102 to ESR 115 ##
 
+### More Versatile `JS_InitClass` ###
+
+In previous versions of SpiderMonkey, the name of the constructor
+function added by `JS_InitClass` was always the name of the prototype
+object's `JSClass`.
+This limitation caused the instances and the prototype to always share
+the same `JSClass` and `JSClassOps` specification.
+If the prototype should behave differently than the instances, custom
+checks were needed.
+Now `JS_InitClass` accepts a `const char *` parameter for naming the
+constructor function, so a different `JSClass *` can be used for the
+prototype object.
+If `nullptr` is passed, the prototype object will be a plain JS object.
+
+### Module Import ###
+
+Previously, `JS::FinishDynamicModuleImport()` when passed an
+`evaluationPromise` argument that was null, would take the pending
+exception from the context and reject the import promise with it.
+This is no longer the case.
+
+If you want the dynamic import promise to reject with a particular
+value, you have to create a rejected promise and pass it to
+`JS::FinishDynamicModuleImport()`.
+If `evaluationPromise` is null, it will reject with a generic value and
+the pending exception will be lost.
+
+An example of how to migrate the code:
+```c++
+// old
+if (JS_IsExceptionPending(cx)) {
+    return JS::FinishDynamicModuleImport(cx, nullptr, referencingPrivate, moduleRequest,
+                                         internalPromise);
+}
+
+// new
+static JSObject* rejectPromiseWithPending(JSContext* cx) {
+    JS::ExceptionStack stack{cx};
+    if (!JS::StealPendingExceptionStack(cx, &stack)) return nullptr;
+    JS::RootedObject rejected{cx, JS::NewPromiseObject(cx, nullptr)};
+    if (!rejected || !JS::RejectPromise(cx, rejected, stack.exception())) return nullptr;
+    return rejected;
+}
+// ...
+if (JS_IsExceptionPending(cx)) {
+    JS::RootedObject rejectedPromise{cx, rejectPromiseWithPending(cx)};
+    return JS::FinishDynamicModuleImport(cx, rejectedPromise, referencingPrivate,
+                                         moduleRequest, internalPromise);
+}
+```
+
+### Bound Function Objects ###
+
+Bound function objects (callable functions created with
+`Function.prototype.bind()`) are no longer `JSFunction` objects.
+`JS::GetBuiltinClass()` will not return `js::ESClass::Function` for such
+objects.
+
+Unless you are using `JS::GetBuiltinClass()` to determine whether an
+object is callable, you will most likely not have to migrate anything.
+Consider using `JS::IsCallable()`.
+
+If you need to check whether an object is specifically a bound function
+object, use `JS_ObjectIsBoundFunction()`.
+
 ### Various API changes ###
 
 This is a non-exhaustive list of minor API changes and renames.
 
 - `JS::ModuleInstantiate` → `JS::ModuleLink`
+- `mozilla::Tuple` → `std::tuple` (`<mozilla/Tuple.h>` is removed, use
+  the standard C++ header `<tuple>` instead)
```

Comment:
Also, it would be nice to mention that the filename used by the engine is UTF-8 ([bug 1492090](https://bugzilla.mozilla.org/show_bug.cgi?id=1492090))

such as `JS::CompileOptions`.

https://searchfox.org/mozilla-esr115/rev/61b47de1faebf23626e519b2464b461589fbea3e/js/public/CompileOptions.h#120,126-130
```cpp
class JS_PUBLIC_API TransitiveCompileOptions {
...
  // UTF-8 encoded file name.
  const char* filename_ = nullptr;

  // UTF-8 encoded introducer file name.
  const char* introducerFilename_ = nullptr;
```